### PR TITLE
environment.py: just pass concrete specs to regenerate

### PR DESCRIPTION
**Pull request by haampie**
Edit: PR was wrong. But now it's just cleaning up the API a bit.

### CODE REVIEW

**Comment by haampie** [comment]
sigh... windows. @johnwparent can you have a look?

**Comment by tgamblin** [comment]
@johnwparent: this is the code that's causing issues:

```
def rename(src, dst):
        # On Windows, os.rename will fail if the destination file already exists
        if is_windows:
            if os.path.exists(dst):
                os.remove(dst)
>       os.rename(src, dst)
```

It looks like the remove failed?  I'm not really sure why it's being hit.  At any rate, On Python 3.3+ there is an `os.replace` call that should probably be used instead.  I think the logic for an atomic move should be stuck into `llnl.util.filesystem` and use the above logic on python 2.7, but `os.replace` on newer pythons.

see https://bugs.python.org/issue8828

**Comment by tgamblin** [comment]
I'm not sure why this PR would cause that to fail though.  So I worry that there is something deeper here.

**Comment by haampie** [comment]
possibly related: https://github.com/spack/spack/issues/29635. but let's focus on getting this PR merged without having to work out hairy details of windows...

File: lib/spack/spack/environment/environment.py [hunk 1]
```diff
@@ -474,14 +474,14 @@ def __contains__(self, spec):
 
         return True
 
-    def specs_for_view(self, concretized_specs):
+    def specs_for_view(self, root_specs):
```
Comment (alalazo) [inline lib/spack/spack/environment/environment.py]: Just a comment. Can we use:
```suggestion
    def specs_for_view(self, concretized_root_specs):
```
here and below? That makes clear that we expect the concretized version of the root specs in these calls (as opposed to the ones that appear in `spack.yaml`)

**Comment by johnwparent** [comment]
> @johnwparent: this is the code that's causing issues:
> 
> ```
> def rename(src, dst):
>         # On Windows, os.rename will fail if the destination file already exists
>         if is_windows:
>             if os.path.exists(dst):
>                 os.remove(dst)
> >       os.rename(src, dst)
> ```
> 
> It looks like the remove failed? I'm not really sure why it's being hit. At any rate, On Python 3.3+ there is an `os.replace` call that should probably be used instead. I think the logic for an atomic move should be stuck into `llnl.util.filesystem` and use the above logic on python 2.7, but `os.replace` on newer pythons.
> 
> see https://bugs.python.org/issue8828

Apologies for the delayed response, but I was having difficulties replicating the issue locally. Now that I have, this is due to the fact that the files being manipulated in this case seems to be directory shortcuts (or links thereof) on Windows, which, as Todd pointed out, is causing the call to remove to fail. This is not directly related to the fact that the files are directories, but rather that the 'dst' file in this case seems to have already been removed, but a reference to it remains in that context.

I'm updating to the paradigm suggested by Todd. @haampie I can add a patch here or push to your branch, which ever is easiest.

**Comment by haampie** [comment]
Can it be a separate PR / commit on develop? Thanks

**Comment by johnwparent** [comment]
> Can it be a separate PR / commit on develop? Thanks

Absolutely! I'll tag you when that's made so you can keep pace with development.

**Comment by haampie** [comment]
Note that (somehow) when all my environments PRs were combined in #29645, Windows tests passed. Not sure if that helps you debugging/fixing, but potentially relevant.

**Comment by johnwparent** [comment]
> Note that (somehow) when all my environments PRs were combined in #29645, Windows tests passed. Not sure if that helps you debugging/fixing, but potentially relevant.

I believe that may be due to some of the view creation error handling you added in one of the amalgamated PRs. Honestly, I think this test was only ever passing by sheer happenstance, and this PR has done a good job of revealing some underlying functional issues on the Windows side of things.

**Comment by haampie** [comment]
Ah, I see what the issue is now. The hash that's used to compute the install dir is does not take into account whether or not these specs are actually installed. So:

1. on concretize spack updates the view
2. on install spack updates the view

Same view hash in (1) and (2)...

So we do have to run full_hash on the *actual* filtered specs added to the view :(. @tgamblin how expensive is it to run full_hash on a list of specs? is there any caching?

**Comment by haampie** [comment]
Okay, simplified it to this:

What we are doing is merging prefixes of effectively selected (based on filters, but also install status) configured through a view descriptor. So the hash can be computed from that: effective prefixes + descriptor in some stable structure. No need for full hashes.

File: lib/spack/spack/environment/environment.py [hunk 2]
```diff
@@ -413,7 +413,7 @@ def _next_root(self, specs):
     def content_hash(self, specs):
         d = syaml.syaml_dict([
             ('descriptor', self.to_dict()),
-            ('specs', [(spec.full_hash(), spec.prefix) for spec in sorted(specs)])
+            ('prefixes', sorted([s.prefix for s in specs]))
```
Comment (becker33) [inline lib/spack/spack/environment/environment.py]: Why remove the spec hashes from this? I know that in any reasonable configuration the prefix includes the hash, but I'm a little worried that if a user removes the hash, they could have different specs in the same prefixes.

Comment (haampie) [reply lib/spack/spack/environment/environment.py]: Okay yeah. Much ado about nothing. Originally I thought the plan was to call `.full_hash()` on the root specs. But I forgot about the install status. Now this PR is just cleaning the API a little bit.

**Review by becker33 — APPROVED** [review]

**Comment by haampie** [comment]
@spackbot run pipeline

**Comment by haampie** [comment]
@spackbot run pipeline

**Comment by haampie** [comment]
@spackbot run pipeline

**Comment by haampie** [comment]
@spackbot run pipeline

**Comment by haampie** [comment]
@spackbot run pipeline

**Comment by haampie** [comment]
@spackbot run pipeline

**Comment by haampie** [comment]
Waiting for pod pipeline/runner-7rhozbqh-project-2-concurrent-0rz9jb to be running, status is Pending
	Unschedulable: "0/20 nodes are available: 20 node(s) didn't match Pod's node affinity/selector."

@scottwittenburg ?

**Comment by haampie** [comment]
@spackbot run pipeline

**Comment by scottwittenburg** [comment]
fyi @opadron @kotfic

**Pull request merged**